### PR TITLE
feat(FR-3205): redesign custom deployment Start Command with shell/exec modes

### LINKED ISSUE #8035: Redesign deployment Start Command input: backend raw-string command/shell field + shell vs exec mode UI
## Background

Deployment **Start Command** bugs (FR-3146 line-continuation, FR-3166 `;`/`&&`/`|`/redirection) share one root cause: the WebUI tokenizes the user's free-text command client-side and sends a bare argv array (`[String!]`), and the manager exec's it with no shell — so shell operators break.

FR-3166 shipped only a **temporary mitigation**: the client still tokenizes; we just added helper text telling users to wrap their command themselves as `/bin/bash -c "cmd1; cmd2"`. The auto-wrap approach was tried and reverted within that PR. The root cause is unchanged.

This issue tracks the **proper design** agreed on the Teams thread.

## Backend prerequisite (separate BA work)

- Deprecate the `[String!]` `start_command`; add a `str` `command` field and an optional `shell` field.
- Value-based dispatch on the backend: `shell` set -> run `shell -c {command`}; `shell` unset -> backend `shlex.split` then exec.
- Default = shell mode, default shell `bash` (fall back to `sh`). All tokenizing/splitting moves to the backend; the WebUI sends the raw string.

The frontend work below can only land once the new `command`/`shell` API exists (version-gated for skew).

## UI design

Wrap the whole command/model configuration region — the `명령어 입력 | 설정 파일 사용` segment down through the port field — into a single **collapsible card**.

- **Card**: collapsible (expand/collapse), reference the vllm-runtime collapsible card in `DeploymentAddRevisionModal.tsx`.
- **Card header extra (right side)**: the existing `명령어 입력 | 설정 파일 사용` Segmented control, moved out of the body into the card header. Each mode gets a tooltip explaining what it does.
- **Basic / Advanced switch**: replaces the former "Advanced" collapse. A switch at the top-right of the command-input area (`기본 | 고급`) with a question-mark tooltip. Reference the session launcher "수동(manual)" switch (switch + question tooltip).
- **Advanced mode** reveals the `Execution` (shell/exec) selector and the `Shell` field, placed **between the Start Command label and its input** (order: label+switch -> Execution -> Shell -> input).
- **Execution-mode-dependent helper & validation**: helper text, validation, and input shape all differ per mode:
  - Basic (shell implicit): no helper.
  - Advanced / Shell: helper "shell operators available (`;` `&&` `|` `$VAR` redirection work)"; multi-line textarea (newlines allowed).
  - Advanced / Exec: warning helper "`;` `&&` `|` `$VAR` redirection NOT available — split into argv"; single-line input (line-wrap allowed for readability, but Enter/newline input blocked); Shell selector disabled; validation warns/blocks on newline or shell operators.

### Wireframes

```
(1) Card expanded — Command input / Basic mode
+- Command / Model config           [ Command input | Config file ] v -+
|                                     (segment in card header extra + tooltip)
|  Start Command (i)                              Basic (o)--( ) Advanced (?)
|  [ vllm serve /models --tp 2                                       ]  <- multiline
|    (no helper)
|  Model folder mount path (i) (optional)
|  [ /models                                                         ]
|  Port (i) (optional)
|  [ 8000                                                            ]
+---------------------------------------------------------------------+

(2) Card collapsed
+- Command / Model config           [ Command input | Config file ] ^ -+
+---------------------------------------------------------------------+

(3) Advanced — Shell   (order: Start Command -> Execution -> Shell -> input)
|  Start Command (i)                              Basic ( )--(o) Advanced (?)
|  Execution   (o) Shell     ( ) Exec
|  Shell       [ bash                              v ]
|  [ chmod -R 777 ...; vllm serve /models ...                        ]  <- multiline
|  helper: shell operators available (; && | $VAR redirection work)
|  Model folder mount path (i) (optional)
|  [ /models                                                         ]
|  Port (i) (optional)
|  [ 8000                                                            ]

(4) Advanced — Exec
|  Start Command (i)                              Basic ( )--(o) Advanced (?)
|  Execution   ( ) Shell     (o) Exec
|  Shell       [ -- disabled -- ]
|  [ vllm serve /models --kv-cache-dtype fp8_e4m3 ...                ]  <- single-line (wrap OK, no newline)
|  (!) ; && | $VAR redirection NOT available — split into argv
|      validation: warn/block on newline or shell operators
|  Model folder mount path (i) (optional)
|  [ /models                                                         ]
|  Port (i) (optional)
|  [ 8000                                                            ]

(5) Config file mode (segment set to right)
+- Command / Model config           [ Command input | Config file ] v -+
|                                     (selected; tooltip: direct YAML)
|  Model folder mount target (i)        Model definition file path (i) (optional)
|  [ /models                    ]       [ model-definition.yaml            ]
|  (no Basic/Advanced switch, no Execution — YAML expresses shell/command directly)
+---------------------------------------------------------------------+
```

## Reference implementations

- Collapsible card + runtime-specific card body: `DeploymentAddRevisionModal.tsx` (vllm runtime card).
- Basic/Advanced switch pattern (switch + question tooltip): session launcher resource config "수동" switch.
- Segment `명령어 입력 | 설정 파일 사용` already exists as `customDefinitionMode: 'command' | 'file'` in `DeploymentAddRevisionModal.tsx`.

## Out of scope

- The YAML / config-file mode (model-definition.yaml) keeps its current fields (mount target + model definition file path) and needs no shell/exec form distinction — YAML carries `shell:`/`command:` directly. The shell/exec distinction applies only to the command (form) mode.

## Affected files

- `react/src/components/DeploymentAddRevisionModal.tsx`
- `react/src/pages/AdminDeploymentPresetSettingPage.tsx` (+ `AdminDeploymentPresetSettingPageContent.tsx`)
- `react/src/helper/parseCliCommand.ts` (remove tokenizer from the write path; keep only for hint extraction)

## Open decisions

- How the backend represents "no shell" (null vs sentinel).
- exec-mode validation strength: warn vs hard-block on newline / shell operators.
- Whether to ship exec mode in v1, or v1 = shell-only and add exec later (most served images include sh, so exec is rarely needed).

## Related

- Follows up FR-3166 (temporary mitigation).

JIRA Issue: FR-3205

### LINKED ISSUE #8323: Drive custom-deployment model-definition placeholders from RuntimeVariant.readsVfolderConfigFiles + defaultModelDefinition
Backend PR lablup/backend.ai#12970 (BA-6942, manager 26.8.0) exposes two new fields on the GraphQL `RuntimeVariant` type: `readsVfolderConfigFiles: Boolean!` and `defaultModelDefinition: RuntimeVariantModelDefinition!`. This resolves the existing `TODO(FR-3205)` in `DeploymentAddRevisionModal.tsx` (~line 1456) that deferred this work until the backend exposed these fields.

**Goal**: replace the hardcoded `variantName === 'custom'` heuristic in the deployment add-revision flow with the semantic `readsVfolderConfigFiles`, and drive the custom command/service placeholders from a two-layer merge.

**Changes**

- Add cheap `readsVfolderConfigFiles` boolean to `BAIRuntimeVariantSelect`'s point-lookup and paginated queries; widen the resolved-variant callback so the modal map carries `{ name, readsVfolderConfigFiles `}.
- Fetch the heavier `defaultModelDefinition` via a modal-owned side query keyed on the selected runtime variant id (parallel to how the vfolder yaml read is keyed on the selected model folder), so it never bloats the shared select payload.
- Placeholder precedence (display-only hints): DB `defaultModelDefinition` (baseline) overridden field-by-field by the vfolder `model-definition.yaml` (kept via the existing client-side read), then the user's own input on top.
- Migrate the `custom` gates (runtime-param preset hide, service-config section show, submit branch, placeholder enablement, warning validator) to `readsVfolderConfigFiles`; leave the modal-mode Preset/Custom toggle name/mode based.
- `command` is a plain `String` (deprecated `start_command` surfaced via alias); surface it as the raw string with NO re-tokenization, consistent with the FR-3205 stop-tokenizing principle.

**UX constraint**: `Model Definition File Path` lives in the Advanced section, below the command/service fields. Its value must NOT feed back into placeholder computation — placeholders are computed only from the DB default plus the fixed-convention vfolder `model-definition.yaml`. Having a lower Advanced field change the hints of fields above it is awkward UX and is explicitly excluded.

**Version gate**: `@since(version: "26.8.0")` on both new fields; no new `client.ts` capability flag (both fields are read with a fallback: `readsVfolderConfigFiles ?? (name === 'custom')`, `defaultModelDefinition ?? null`). Pin to the first real pre-release tag string if the field ships in an rc before GA (PEP440 rc < final trap).

**Blocked on**: `data/schema.graphql` must be synced with #12970's new `RuntimeVariant` fields before `pnpm relay` compiles. Ship as a stacked follow-up on top of the FR-3205 PR.

**Out of scope**: admin deployment preset command/model-path placeholder hints (separate follow-up); extending `ParsedModelDefinition` with interval/maxWaitTime/expectedStatusCode health-check hints (optional trailing scope).

JIRA Issue: FR-3342

### PR DESCRIPTION
**Pull request by agatha197**
Resolves #8035 (FR-3205)
Resolves #8323 (FR-3342)

> [!WARNING]
> It is blocked by BA-6613, BA-7210

## Summary

Two related changes to the custom deployment revision flow, plus a restructuring
of the admin deployment preset form to unify the service configuration UI.

**FR-3205 — Start Command redesign.** The WebUI stops tokenizing the command
client-side and sends the raw string via the `command`/`shell` fields on
`ModelServiceConfigInput`, letting the backend either wrap it (`shell -c cmd`)
or `shlex.split` + exec. Follow-up to #7957 (FR-3166), which only shipped a
helper-text mitigation while the client kept tokenizing.

**FR-3342 — runtime-variant-driven model definition.** `RuntimeVariant` gained
`readsVfolderConfigFiles` and `defaultModelDefinition`. Whether a variant reads
its config from the mounted vfolder now decides what the form shows, and the
variant's built-in definition supplies the placeholder hints.

### Add Revision Modal (Custom mode)

**The `Enter Command` / `Use Config File` segmented toggle is gone.** Both now
live in one form and the choice is implied by what you fill in.

The command controls sit in one **collapsible card** (Service Configuration):

- **Basic**: shell implicit, multi-line input, no helper text.
- **Advanced → Shell**: reveals `Execution (Shell | Exec)` + `Shell` selector.
- **Advanced → Exec**: single-line input, Shell disabled, sends `shell: null`.
- The Start Command is **optional** — an empty command submits.

**Always visible (regardless of runtime variant):**
- **Health Check** — checkbox + 6 config fields (path, interval, max retries, max wait time, expected status code, initial delay)
- **Pre-Start Actions** — dynamic list of action name + JSON args pairs (newly added)

Driven by `readsVfolderConfigFiles`:
- **Reads config files** → Service Configuration section + Model Definition File Path shown
- **Does not** → both hidden, runtime-parameter presets shown instead

### Admin Deployment Presets

**Restructured step layout:**

**Step 1 (Basic Info)** now includes:
- Name, Description, Runtime Variant, Image (unchanged)
- **Service Configuration** (port, command, shell) — shown only when custom variant selected, with identical UI to the revision modal (Collapse with Basic/Advanced Segmented, Execution Radio, Shell selector)
- **Health Check** — always visible, checkbox + 6 fields
- **Pre-Start Actions** — always visible, dynamic action/args list
- Resources, Deployment Defaults (unchanged)

**Step 2 (Model & Execution):**
- Startup Command, Bootstrap Script, Env Vars (unchanged)
- **Model Definition card simplified** — only model name, model path, and metadata. The former "Configure Service" inner card (port, command, shell, pre-start actions, health check) has been removed; those fields now live in Step 1.

**Step 3 (Review):**
- Service config (port, command, shell, health check, pre-start actions) shown in the **Basic Info** review card with Edit → Step 1
- Model Definition review shows only name, path, and metadata

**Submit handler:** Service config fields are now **decoupled from the model definition switch**. Port, command, health check, and pre-start actions are submitted even when the model definition switch is off; only model name/path/metadata are gated by the switch.

### Shared changes

- **"Add Metadata" → "Metadata"** (i18n label change)
- **"Start Command" → "Command"** when `supportsCommandShell` (26.8.0+); legacy managers keep "Start Command"
- **Exec mode label**: "Command (argv)" to distinguish from shell commands
- Added `modelService.Command` i18n key (en: "Command", ko: "명령어")
- Added `modelService.PreStartActionsTooltip` with tooltips on both pages
- Port field uses `modelService.PortTooltip`

## Version gating

The command/shell path is gated at **26.8.0**, not at the 26.7.0 tag that first
carried the fields. The GraphQL input default `shell: String = "/bin/bash"`
(BA-6742) merged *after* that tag, so on a 26.7.0 manager an omitted `shell` is
null — which disables shell wrapping and silently turns Basic mode into Exec.

Capability is read through the client feature flags
(`model-service-command-string`, `model-runtime-variant-reads-vfolder-config-files`)
rather than inline version comparisons.

## Files

- `react/src/helper/modelServiceCommand.ts` — shared prefill/submit logic
- `react/src/helper/parseModelDefinitionYaml.ts` — partial YAML parse + GraphQL default normalization
- `react/src/hooks/useModelDefinitionDefaults.ts` — reads selected folder's `model-definition.yaml`
- `packages/backend.ai-ui/src/components/fragments/BAIRuntimeVariantSelect.tsx` — resolves variant metadata
- `packages/backend.ai-client/src/client.ts` — feature flags
- `AdminDeploymentPresetSettingPageContent.tsx` — restructured step layout
- `AdminDeploymentPresetModelConfigItem.tsx` — simplified to name/path/metadata only
- `AdminDeploymentPresetReviewSummary.tsx` — service config in Basic Info card
- `AdminDeploymentPresetSettingPage.tsx` — submit handler decoupled from model def switch
- `AdminDeploymentPresetFormTypes.ts` — added `commandExecution` field
- `DeploymentAddRevisionModal.tsx` — pre-start actions, command label, port-only submit
- `resources/i18n/*.json` — en + ko translated; other locales English fallback

## Verification

- `bash scripts/verify.sh` → `=== ALL PASS ===`
- Relay, TypeScript, Lint, Format, Terminology all pass

---

### Astryx rebase note (2026-08-11)

Rebased onto main after the Ant Design → Astryx migration (#8626). All UI in this PR was re-expressed in the Astryx design system (form-engine `Form`/`BAIFormItem`, `AstryxForm*` controls, `Collapsible`, `MetadataList`, tokens-only styling) with feature semantics unchanged — the pure-TS layer (`modelServiceCommand.ts`, submit mapping, capability gates) is byte-identical to the pre-rebase branch. Also: restored main's `data/schema.graphql` snapshot (the branch's copy silently reverted ~220 lines of unrelated main types; only the intended `PresetModelConfigInput.name/modelPath` nullable change remains), moved the Basic/Advanced toggle out of the Collapsible trigger (Astryx renders triggers inside a `<button>`, which cannot host interactive controls), and dropped the removed 'Use Config File' toggle interactions from two main e2e specs this PR's UI removal had broken. Verified: `scripts/verify.sh` ALL PASS, react vitest 1174 passed, BUI vitest 583 passed.

### CODE REVIEW

**Comment by agatha197** [comment]
* **#8170** <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/8170?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#8324](https://github.com/lablup/backend.ai-webui/pull/8324) <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/8324?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#8333](https://github.com/lablup/backend.ai-webui/pull/8333) <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/8333?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈 <a href="https://app.graphite.com/github/pr/lablup/backend.ai-webui/8170?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>flow:merge-queue</kbd> - adds this PR to the back of the merge queue
- <kbd>flow:hotfix</kbd> - for urgent changes, fast-track this PR to the front of the merge queue

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/lablup?ref=merge-queue-instructions-in-stack-comment&prId=8170).

<h2></h2>
<sub>

An organization admin has required the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: resources/i18n/en.json [hunk 1]
```diff
@@ -1934,13 +1935,22 @@
     "AdditionalMountsTooltip": "Extra non-model folders mounted into each replica alongside the model folder.",
     "AutoScalingRules": "Auto Scaling Rules",
     "BasePath": "Base Path",
+    "CommandAdvancedModeTooltip": "Advanced mode lets you choose how the command is executed (Shell or Exec) and which shell to use.",
+    "CommandExecHelper": "Shell operators (; && | $VAR, redirection, etc.) cannot be used. Enter each argument separated by spaces.",
+    "CommandExecNewlineError": "The command must be a single line in Exec mode.",
```
[the remark this answers is not in the document]
Comment (agatha197) [reply resources/i18n/en.json]: Good catch — removed the orphaned `CommandExecNewlineError` key from all locale files. Exec mode uses a single-line `<Input>` (no newline possible), so no newline validator is needed; the only wired validator is the shell-operator check (`CommandExecOperatorWarning`).

File: react/src/helper/modelServiceCommand.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+ *   Basic, any other value → Advanced/Shell with that shell.
+ * - When only the deprecated `startCommand` token list exists (legacy
+ *   revisions, or managers that strip `command`), reconstruct the command
+ *   string via `formatShellCommand` and fall back to Basic mode.
+ */
+export function deriveCommandModeState(params: {
+  command?: string | null;
+  shell?: string | null;
+  startCommand?: readonly string[] | null;
+}): CommandModeState {
```
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/helper/modelServiceCommand.ts:67]: Added `react/src/helper/modelServiceCommand.test.ts` (11 cases) covering the mode-derivation edge cases: null shell → Exec, default `/bin/bash` → Basic, custom shell → Advanced/Shell, empty/absent `command` → fallback to `startCommand`, and `resolveCommandShell` for Basic/Shell/Exec + the preset (omitted-execution) path.

**Review by nowgnuesLee — CHANGES_REQUESTED** [review]
The other locale files all seem to contain English text — please double-check.

**Review by ironAiken2 — DISMISSED** [review]
please resolve the conflicts

File: react/src/components/DeploymentAddRevisionModal.tsx [hunk 3]
```diff
@@ -1726,89 +1856,304 @@ const DeploymentAddRevisionModal: React.FC<DeploymentAddRevisionModalProps> = ({
...
+                      // keeps switching modes from toggling the collapse.
+                      label: (
+                        <BAIFlex
+                          justify="between"
+                          align="center"
+                          gap="sm"
+                          style={{ flex: 1 }}
+                        >
+                          <span>{t('modelService.ServiceConfiguration')}</span>
+                          {supportsCommandShell && (
```
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/components/DeploymentAddRevisionModal.tsx]: By design — the `Command input | Config file` mode and the `definitionPath` field were intentionally removed (product decision to discourage creating revisions from a hand-written `model-definition.yaml`; the model folder’s definition is read from the folder instead). The old wireframe/PR contract is superseded, and a command + definition path can no longer be submitted together. The PR description will be updated to reflect the new scope.

File: react/src/helper/modelServiceCommand.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  const { command, shell, startCommand } = params;
+  const usingNewCommand = command != null && command !== '';
+  const commandString = usingNewCommand
+    ? command!
+    : formatShellCommand(startCommand ?? []);
+
+  if (!usingNewCommand) {
+    return {
+      command: commandString,
+      advanced: false,
```
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/helper/modelServiceCommand.ts]: Fixed. A legacy `startCommand` argv is now reconstructed as **Exec** (shell submitted as `null`) so the tokens round-trip via `shlex.split` without shell interpretation. Reconstructing it as Shell would wrap the string and re-interpret `;` / expand `$VAR`, changing semantics. (react/src/helper/modelServiceCommand.ts)

File: react/src/helper/modelServiceCommand.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      execution: 'shell',
+      shell: DEFAULT_MODEL_SERVICE_SHELL,
+    };
+  }
+  if (shell == null) {
```
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/helper/modelServiceCommand.ts]: Fixed. `deriveCommandModeState` now treats both `null` and empty/whitespace `shell` as Exec (no shell wrapping), matching the backend contract, so a stored `shell: ""` round-trips as Advanced/Exec instead of Advanced/Shell with an empty required field.

File: react/src/helper/modelServiceCommand.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+];
+
+export type CommandExecutionMode = 'shell' | 'exec';
+
+/**
+ * Shell operators / expansions that only work under a shell (Execution = Shell).
+ * Used to warn when they appear in an Exec-mode command, where they are passed
+ * literally to `execve` and never interpreted.
+ */
+export const SHELL_OPERATOR_PATTERN = /[;&|<>$`]/;
```
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/helper/modelServiceCommand.ts]: Acknowledged — the character-only check can false-positive on operators inside quoted argv values (e.g. `python -c "print(1 > 0)"`). The hard block is a deliberate product choice here (prefer blocking over a silent warning). The quoted/escaped false-positive is a known limitation; making it shlex/quote-aware is a possible follow-up. Leaving open for the maintainer to weigh block-vs-warn.

File: react/src/hooks/useModelDefinitionDefaults.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        const content = await (await response.blob()).text();
+        return parseModelDefinitionYaml(content);
+      } catch {
+        return null;
+      }
+    },
+    enabled: enabled && !!vfolderId,
+    // Cache per vfolder for the modal's lifetime so re-renders and mode
+    // toggles don't re-download the same definition file.
+    staleTime: Infinity,
```
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/hooks/useModelDefinitionDefaults.ts]: Acknowledged — `staleTime: Infinity` will not pick up edits to `model-definition.yaml` on revisit. This placeholder-from-model-definition feature is currently under product reconsideration; if it stays, the query will use a finite staleTime / revalidation and drop inactive entries. Leaving open pending that decision.

File: packages/backend.ai-client/src/client.ts [hunk 8]
```diff
@@ -933,6 +933,19 @@ export class Client {
       // TODO(FR-3139): simplify to '26.4.4' once rc builds are out of use.
       this._features['model-runtime-variant-preset-values'] = true;
     }
+    if (this.isManagerVersionCompatibleWith('26.4.4')) {
+      // ModelMountConfigInput / ExtraVFolderMountInput gained `subpath` in
+      // 26.4.4 (BA-6242): mount a subfolder inside the model vfolder instead of
+      // its root. Older managers reject the unknown input field, so the key is
+      // omitted from the mutation entirely on them.
```
[the remark this answers is not in the document]
Comment (agatha197) [reply packages/backend.ai-client/src/client.ts]: Acknowledged — the model-mount `subpath` was added as part of the model-folder mount rework in this PR. The final PR title/description will be updated to reflect the expanded scope (Start Command redesign + mount destination/subpath). Not splitting into a separate PR at this time.

File: react/src/helper/modelServiceCommand.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/helper/modelServiceCommand.ts]: Fixed in ef99cdc9a. Removed the Exec-mode operator validator entirely (and the now-dead `SHELL_OPERATOR_PATTERN` / `CommandExecOperatorWarning`). The write path already sends the user's raw command string as-is, so pre-validating operators contradicted that goal and rejected valid quoted argv like `python -c "print(1 > 0)"`. The Exec helper text still explains that unquoted operators are not interpreted, so the guidance stays without the false-positive block.

File: react/src/hooks/useModelDefinitionDefaults.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/hooks/useModelDefinitionDefaults.ts]: Fixed in ef99cdc9a. Changed `staleTime: Infinity` to `staleTime: 0` so an edited `model-definition.yaml` is revalidated on the next modal open. React Query still only refetches on mount / vfolder-key change, so re-renders and mode toggles keep reusing the cached parse — this stays a single lightweight read per open rather than a per-render download.

File: react/src/pages/AdminDeploymentPresetSettingPage.tsx [hunk 9]
```diff
@@ -52,13 +59,27 @@ const buildModelDefinitionInput = (
...
+                  command: service.startCommand,
+                  // Presets always run under a shell (input `shell` is
+                  // non-null), so pass Shell explicitly (no Exec mode) — the
+                  // `??` only satisfies the non-null type, never triggers.
+                  shell:
+                    resolveCommandShell({
+                      advanced: !!service.commandAdvanced,
+                      execution: 'shell',
+                      shell: service.shell,
+                    }) ?? DEFAULT_MODEL_SERVICE_SHELL,
```
Comment (nowgnuesLee) [inline react/src/pages/AdminDeploymentPresetSettingPage.tsx]: Why is the fallback value DEFAULT_MODEL_SERVICE_SHELL instead of undefined?

**Review by nowgnuesLee — DISMISSED** [review]

File: react/src/pages/AdminDeploymentPresetSettingPage.tsx (hunk 9 above)
Comment (agatha197) [reply react/src/pages/AdminDeploymentPresetSettingPage.tsx]: Follow-up: it is now `undefined`, as you suggested — the gate moved instead.

My previous reply argued for sending `/bin/bash` explicitly because BA-6742 (the input default) is unreleased. That was treating the symptom. The cleaner fix is to stop using the `command`/`shell` path on managers that lack the default at all:

- `model-service-command-string` now gates at **26.8.0** instead of 26.7.0.
- BA-6742 and the 26.8.0 RuntimeVariant fields (BA-6942, lablup/backend.ai#12970) are both already in `main`, so they ship together — every manager that takes the command/shell path will carry `shell: String = "/bin/bash"`.
- 26.7.0 managers fall back to the deprecated `startCommand` token list and never receive `shell`, so the null-shell hazard disappears rather than being worked around.
- `resolveCommandShell` therefore returns `undefined` in Basic mode again, and the `?? DEFAULT_MODEL_SERVICE_SHELL` at this call site stays removed.

The cost is that 26.7.0 managers do not get the new Shell/Exec modes and keep client-side tokenization. Given 26.7.0 has no usable server default, enabling the path there would have required the client to hard-code a shell anyway, so gating is the more honest boundary — but flag it if you would rather keep 26.7.0 on the new path.

Note the `@since(version: "26.7.0")` directives on `command` / `shell` are deliberately left alone: those mirror the schema, where the fields really were added in 26.7.0. The 26.8.0 boundary is a client policy and lives in the capability flag, and the comments now keep the two separate.

Still leaving this unresolved for your review.

File: react/src/pages/AdminDeploymentPresetSettingPage.tsx [hunk 10]
```diff
@@ -281,6 +333,15 @@ const AdminDeploymentPresetSettingPage: React.FC = () => {
...
     const presetValues = collectRuntimePresetValuesRef.current();
 
+    // Resolve whether the selected variant reads vfolder config files so
+    // buildModelDefinitionInput can exclude stale command/port data.
+    const selectedVariant = runtimeVariantList.find(
+      (rt) => toLocalId(rt.id) === values.runtimeVariantId,
+    );
+    const reads =
+      selectedVariant?.readsVfolderConfigFiles ??
+      selectedVariant?.name === 'custom';
```
[the remark this answers is not in the document]
Comment (agatha197) [reply react/src/pages/AdminDeploymentPresetSettingPage.tsx:408]: Already addressed in d12739885 — this thread was left open by mistake.

`AdminDeploymentPresetSettingPage.tsx` now runs a point lookup for the selected variant (`selectedRuntimeVariantLookup`) and appends it to the paginated list when it falls outside the 100-item page:

```ts
const runtimeVariantList =
  selectedRuntimeVariantLookup &&
  !paginatedRuntimeVariantList.some((rt) => rt.id === selectedRuntimeVariantLookup.id)
    ? [...paginatedRuntimeVariantList, selectedRuntimeVariantLookup]
    : paginatedRuntimeVariantList;
```

Because the merge happens on the list itself rather than at the call site, every consumer resolves the selected variant consistently — both the Runtime Select's options and the `readsVfolderConfigFiles` derivation at submit time — so editing a preset whose variant is outside the page no longer drops its command/port fields.

File: packages/backend.ai-ui/src/components/fragments/BAIRuntimeVariantSelectAstryx.tsx [hunk 11]
```diff
@@ -37,19 +37,26 @@ export interface BAIRuntimeVariantSelectProps extends Omit<
...
+   * elsewhere in the form (e.g., to branch on whether the variant reads the
+   * vfolder config files) without re-querying.
    */
-  onResolvedNamesChange?: (nameMap: Record<string, string>) => void;
+  onResolvedVariantsChange?: (
+    variantMap: Record<
+      string,
+      { name: string; readsVfolderConfigFiles: boolean }
+    >,
+  ) => void;
```
[the remark this answers is not in the document]
Comment (agatha197) [reply packages/backend.ai-ui/src/components/fragments/BAIRuntimeVariantSelectAstryx.tsx:76]: Already addressed in d12739885 — this thread was left open by mistake.

`onResolvedNamesChange` is restored on `BAIRuntimeVariantSelectAstryx` as a `@deprecated` prop rather than removed, so existing backend.ai-ui consumers keep compiling on upgrade:

```ts
/**
 * @deprecated Use `onResolvedVariantsChange` instead — this only surfaces
 * `name`, not `readsVfolderConfigFiles`. Kept (and still notified
 * alongside the richer callback) so existing consumers of this shared
 * backend.ai-ui component don't break on upgrade.
 */
onResolvedNamesChange?: (nameMap: Record<string, string>) => void;
```

Both callbacks are notified during the migration — the effect bails only when neither is supplied, and the deprecated one receives `_.mapValues(variantMap, (v) => v.name)` derived from the same resolved map.

**Review by yomybaby — APPROVED** [review]
LGTM.

Walked the whole stack in execution order (#8170 → #8603 → #8333 → #8622). The `command` + `shell` contract is well documented, and the capability comment explaining *why* the gate sits at 26.8.0 rather than 26.7.0 — the BA-6742 default landing after the 26.7.0 tag, so version strings cannot tell the two apart — is the single most valuable thing in this PR. That is not recoverable from the code.

Conflict with `main` is resolved: merged `main` into `fr-3205` (fast-forward push, no history rewrite). The only content conflict was in `react/src/components/FolderExplorerModalV2.test.tsx`, where both sides had independently added the same `useResolveImageReference` stub — the code line auto-merged and only the comment above it conflicted. Both rationales are kept, condensed to three lines. `bash scripts/verify.sh` → ALL PASS. #8603/#8333/#8622 stay conflict-free against the new tip (verified with `git merge-tree`).

**One non-blocking note, needs a decision rather than a change:**

The new gates (`model-service-command-string`, `model-runtime-variant-reads-vfolder-config-files`, both `>= 26.8.0`) have no rc pin, so they stay shut on any manager built from an rc tag.

Measured on our staging manager:

- `GET http://10.82.0.130:8090/func/` → `{"version": "v9.20250722", "manager": "26.8.0rc1"}`
- `comparePEP440Versions('26.8.0rc1', '26.8.0') === -1` (PEP 440 sorts a release candidate *before* its release)

So on that manager the Execution/Shell controls never render and the legacy `startCommand` path is what QA sees. There is precedent for handling this in the same file — `model-health-check-enable` is pinned to `'26.4.4rc7'` with a comment saying "Pinned to the rc7 tag so the flag also activates against staging managers built from that tag", plus a `TODO` to simplify once the final release ships.

Is the intent to wait for the 26.8.0 final on staging, or should these gates get the same rc pin so the path is reachable before then? Not a code defect either way — just a question of whether this stack is testable where the team actually tests it.

FYI: @nowgnuesLee's `CHANGES_REQUESTED` from 2026-07-28 is still outstanding and my approval does not clear it.

**Comment by yomybaby** [comment]
@nowgnuesLee — following up on your 2026-07-10 review:

> The other locale files all seem to contain English text — please double-check.

That was left in the review body rather than an inline thread, so GitHub had nothing to mark resolved and it was never answered. It is still true on the current tip. Checked across the 20 non-English locales:

| Key | Status |
|---|---|
| `modelService.Subpath` | English in **19** locales (only `ko` translated) |
| `modelService.SubpathTooltip` | English in **19** locales — a full sentence, so it surfaced verbatim in the UI |
| `modelService.ExecutionExec` / `ExecutionShell` | English in all 20, but these are `Exec` / `Shell` — technical terms the project deliberately leaves untranslated, so left alone |

Rather than reopen this PR after it has been reviewed and approved, the translations ship in **#8973**, stacked on top of #8622 at the end of this stack. Each one reuses the folder/mount vocabulary already established in that locale (`general.Folder`, `session.launcher.FolderToMount`) instead of introducing new terms.

Your other point — the `DEFAULT_MODEL_SERVICE_SHELL` fallback at `AdminDeploymentPresetSettingPage.tsx:80` — was addressed in this PR by moving the capability gate to 26.8.0 rather than patching the fallback, so that one is resolved here.

Flagging this because I approved this PR earlier today on the basis that its unresolved-thread count was zero. That count does not see findings left in a review body, and this one was live. Sorry for the noise.
